File: other/examples-java/src/main/java/com/joshcough/minecraft/betterexamples/MultiPlayerCommands.java
package com.joshcough.minecraft.betterexamples;

import com.joshcough.minecraft.betterjava.*;
import static com.joshcough.minecraft.betterjava.JavaParsers.*;

import org.bukkit.GameMode;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;

public class MultiPlayerCommands extends BetterJavaPlugin {

  public MultiPlayerCommands(){

    commands.add(Command("goto", "Teleport to a player.",
      either(player, integer.and(integer).and(opt(integer))),
        (you, e) -> {
          if(e.isLeft()) you.teleport(e.getLeft());
          else{
            Tuple2<Tuple2<Integer, Integer>, Option<Integer>> t = e.getRight();
            int x = t._1()._1();
            int z = t._1()._2();
            t._2().foldV(
              () -> teleport(you, x, you.getWorld().getHighestBlockYAt(x, z), z),
              y  -> teleport(you, x, z, y)
            );
          }
        }
    ));

    commands.add(Command("set-time", "Sets the time.", integer,
      (p,i) -> p.getWorld().setTime(i.longValue())
    ));

    commands.add(Command("day", "Sets the time to day.", p -> p.getWorld().setTime(1)));

    commands.add(Command("night", "Sets the time to night.", p -> p.getWorld().setTime(15000)));

    commands.add(Command("entities", "Display all the entities.", p -> {
      for(Entity e: p.getWorld().getEntities()){ p.sendMessage(e.toString()); }
    }));

    commands.add(Command("feed", "Fill a players hunger bar", player,
      (you, them) -> doTo(you, them, () -> them.setFoodLevel(20), "fed")
    ));

    commands.add(Command("starve", "Drain a players hunger bar", player,
      (you, them) -> doTo(you, them, () -> them.setFoodLevel(0), "starved")
    ));

    commands.add(Command("shock", "Shock a player.", player,
      (you, them) -> doTo(you, them, () -> them.getWorld().strikeLightning(them.getLocation()), "starved")
    ));

    commands.add(Command("gm", "Set your game mode", gamemode, (you, gm) -> you.setGameMode(gm)));

    commands.add(Command("gms", "Set your game mode to survival", you -> you.setGameMode(GameMode.SURVIVAL)));

    commands.add(Command("gmc", "Set your game mode to creative", you -> you.setGameMode(GameMode.CREATIVE)));

    commands.add(Command("spawn", "Spawn some mobs.", entity.and(opt(integer)),
      (p, t) -> spawnN(t._1(), t._2().getOrElse(() -> 1), p.getLocation())
    ));

    commands.add(Command("ban", "Ban some players.", anyString.plus(),
      (you, ss) -> {
        for(Player them: getServer().getOnlinePlayers()){
          if(ss.contains(them.getName())) ban(them, you.getName() + " doesn't like you.");
        }
        for(OfflinePlayer them: getServer().getOfflinePlayers()){
          if(ss.contains(them.getName())) them.setBanned(true);
        }
      }
    ));

    commands.add(Command("up", "Go up to the surface.", p ->
      p.teleport(p.getWorld().getHighestBlockAt(p.getLocation()).getLocation())
    ));

    commands.add(Command("kill", "Kill entities.", either(match("player").and(player), entity),
      (killer, e) -> {
          if(e.isLeft()) doTo(killer, e.getLeft()._2(), () -> e.getLeft()._2().setHealth(0), "killed");
          else for(Entity en: killer.getWorld().getEntities()){
            if(en.getType() == e.getRight()) en.remove();
          }
      }
    ));
//
//  /**
//   TODO: still need to implement these, but they require all the block pimping
//   which will be non-trivial and ugly in java.
//
//   Command("box",      "Put a box around yourself, made of any material.",
//     args(material){ case p ~ m  => p.blocksAround.foreach(_ changeTo m) }),
//
//   Command("safe",     "Put yourself in a box made of bedrock.",
//    noArgs(_.blocksAround.foreach(_ changeTo BEDROCK))),
//
//   Command("drill",    "Drill down to bedrock immediately.", noArgs(p =>
//     for (b <- p.blockOn.blocksBelow.takeWhile(_ isNot BEDROCK); if (b isNot AIR)) {
//     b.erase
//     if (b.blockBelow is BEDROCK) b.nthBlockAbove(2) changeTo STATIONARY_WATER
//    })),
//
//   Command("creeper-kill", "Surround a player with creepers", opOnly(p2p((_, them) => {
//    them.setGameMode(SURVIVAL)
//    them.loc.block.neighbors8.foreach(_.loc.spawn(CREEPER))
//   })
//   **/
  }
}

File: examples/MultiPlayerCommands/MultiPlayerCommands.scala
package com.joshcough.minecraft.examples

import org.bukkit.GameMode._
import org.bukkit.Material._
import org.bukkit.entity.EntityType._
import scala.collection.JavaConversions._
import com.joshcough.minecraft.CommandsPlugin
import org.bukkit.Material

/**
 * Classic MultiPlayerCommands plugin, done in Scala.
 * Gives a whole pile of useful commands.
 * Their descriptions serve well as documentation.
 */
class MultiPlayerCommands extends CommandsPlugin {

  val commands = List(

    Command("goto",     "Teleport!", player or location){ case (you, e) =>
      e.fold(them => you teleportTo them, loc => you.teleport(loc of you.world))
    },

    Command("up",       "Go up to the surface.")(_.surface),

    Command("set-time", "Sets the time.", time){ case (p, n) => p.world setTime n },

    Command("day",      "Sets the time to day (1000).")   (_.world setTime 1000),

    Command("night",    "Sets the time to night (15000).")(_.world setTime 15000),

    Command("gms",      "Set your game mode to survival.")(_ setGameMode SURVIVAL),

    Command("gmc",      "Set your game mode to creative.")(_ setGameMode CREATIVE),

    Command("gm",       "Set your game mode.", gamemode){ case (p, gm) => p setGameMode gm },

    Command("entities", "Display all the entities.")(p => p !* (p.world.entities.map(_.toString): _*)),

    Command("spawn",    "Spawn some mobs.", entity ~ int.?.named("number to spawn")){
      case (p, e ~ n) => p.loc.spawnN(e, n.fold(1)(id))
    },

    // Can now do this this way from https://www.spigotmc.org/threads/banning-players.131889/ :
    // Bukkit.getBanlist(Type.NAME).addBan(username, reason, expires, source);
//    Command("ban",      "Ban some players.", anyString.named("player").+){ case (you, them) =>
//      server.findOnlinePlayers (them) foreach { _ ban s"${you.name} doesn't like you." }
//      server.findOfflinePlayers(them) foreach { _ setBanned true }
//    },

    Command("box",      "Put a box around yourself, made of any material.", material){ case (p,m) =>
      p.blocksAround.foreach(_ changeTo m)
    },

    Command("safe",      "Put yourself in a box made of bedrock.")(
      _.blocksAround.foreach(_ changeTo BEDROCK)
    ),

    OpOnly(P2P("punish", "Puts someone in a box made of bedrock.")((you, them) =>
      you.doTo(them, them.blocksAround.foreach(_ changeTo BEDROCK), "punished")
    )),

    Command("drill",    "Drill down to bedrock immediately.")(p =>
      for (b <- p.blockOn.blocksBelow.takeWhile(_ isNot BEDROCK); if b isNot AIR) {
        b.erase
        if (b.blockBelow is BEDROCK) b nthBlockAbove 2 changeTo WATER
      }
    ),

    Command("kill",     "Kill entities.", ("player" ~ player) or entity){
      case (killer, Left(_ ~ deadMan)) => killer kill deadMan
      case (killer, Right(e)) => killer.world.entities filter { _ isAn e } foreach (_.remove)
    },

    OpOnly(P2P("creeper-kill", "Surround a player with creepers")((_, them) => {
      them setGameMode SURVIVAL
      them.loc.block.neighbors8 foreach (_.loc spawn CREEPER)
    })),

    OpOnly(P2P("feed",     "Fill a players hunger bar.")((you, them) =>
      you.doTo(them, them setFoodLevel 20, "fed"))
    ),

    OpOnly(P2P("starve",   "Drain a players hunger bar.")((you, them) =>
      you.doTo(them, them setFoodLevel 0, "fed"))
    ),

    OpOnly(P2P("shock",    "Shock a player.")((you, them) => you.doTo(them, them.shock, "shocked"))),

//    OpOnly(Command("lamp", "Spawns a lamp wherever you are looking."){ p =>
//      val b = p.getTargetBlock(null, 1000)
//      if((b is Material.AIR) || math.abs(p.x - b.x) > 50 || math.abs(p.z - b.z) > 50)
//        p ! s"That's more than 50 blocks away. (x: ${math.abs(p.x - b.x)}) (z: ${math.abs(p.z - b.z)})"
//      else {
//        b.blockBelow.changeTo(Material.REDSTONE_TORCH_ON)
//        b.changeTo(Material.REDSTONE_LAMP_ON)
//      }
//    }),

    Command("jump", "Jump really high.", (int ~ int ~ int) or eof){
      case (p, Left(x ~ y ~ z)) => p.setVelocity(new org.bukkit.util.Vector(x, y, z))
      case (p, Right(_))        => p.setVelocity(new org.bukkit.util.Vector(0, 10, 0))
    }
  )
}

/**

p = player
t = target

looking down x axis:
  p.x = 0, t.x = 10, p.z = 0, t.z = 0
  put all the weight in x: Vector(x=10, y=10, z=0)

looking down z axis:
  p.z = 0, t.z = 10, p.x = 0, t.x = 0
  put all the weight in z: Vector(x=0, y=10, z=10)

looking in between
  p.x = 0, p.z = 10, t.x = 5, t.z = 5
  put equal weight in both: Vector(x=10, y=10, z=10)

looking more towards the x axis:
  p.x = 0, p.z = 10, t.x = 7.5, t.z = 2.5
  hmm...use those weights?:  Vector(x=7.5, y=10, z=2.5)

similar with z axis.

*/